sdk/openai/Azure.AI.OpenAI/src/Custom/SseReader.cs

```diff
@@ -0,0 +1,147 @@
+    public sealed class SseReader : IDisposable
+    {
+        private readonly Stream _stream;
+        private readonly StreamReader _reader;
+        private bool _disposedValue;
+
+        public SseReader(Stream stream)
+        {
+            _stream = stream;
+            _reader = new StreamReader(stream);
+        }
+
+        public SseLine? TryReadSingleFieldEvent()
+        {
+            while (true)
+            {
+                var line = TryReadLine();
+                if (line == null)
+                    return null;
+                if (line.Value.IsEmpty)
+                    throw new InvalidDataException("event expected.");
+                var empty = TryReadLine();
+                if (empty != null && !empty.Value.IsEmpty)
+                    throw new NotSupportedException("Multi-filed events not supported.");
+                if (!line.Value.IsComment)
+                    return line; // skip comment lines
+            }
+        }
+
+        // TODO: we should support cancellation tokens, but StreamReader does not in NS2
+        public async Task<SseLine?> TryReadSingleFieldEventAsync()
+        {
+            while (true)
+            {
+                var line = await TryReadLineAsync().ConfigureAwait(false);
+                if (line == null)
+                    return null;
+                if (line.Value.IsEmpty)
+                    throw new InvalidDataException("event expected.");
+                var empty = await TryReadLineAsync().ConfigureAwait(false);
+                if (empty != null && !empty.Value.IsEmpty)
+                    throw new NotSupportedException("Multi-filed events not supported.");
+                if (!line.Value.IsComment)
+                    return line; // skip comment lines
+            }
+        }
+
+        public SseLine? TryReadLine()
+        {
+            if (_reader.EndOfStream)
+                return null;
+            string lineText = _reader.ReadLine();
+            if (lineText.Length == 0)
+                return SseLine.Empty;
+            if (TryParseLine(lineText, out SseLine line))
+                return line;
+            return null;
+        }
+
+        // TODO: we should support cancellation tokens, but StreamReader does not in NS2
+        public async Task<SseLine?> TryReadLineAsync()
+        {
+            if (_reader.EndOfStream)
+                return null;
+            string lineText = await _reader.ReadLineAsync().ConfigureAwait(false);
+            if (lineText.Length == 0)
+                return SseLine.Empty;
+            if (TryParseLine(lineText, out SseLine line))
+                return line;
+            return null;
+        }
+
+        private static bool TryParseLine(string lineText, out SseLine line)
+        {
+            if (lineText.Length == 0)
+            {
+                line = default;
+                return false;
+            }
+
+            int colonIndex = lineText.IndexOf(':');
```

Comment:
I am pretty sure lineText.AsSpan().IndexOf is faster here.

Reply:
OK, I have to confess that my initial reaction to that was "what? no way that can be true...". But I think you're completely right!
https://dotnetfiddle.net/PCc1ct

Using the span consistently looks to shave off at least 10-15% of the runtime. Pragmatically, the difference here is immaterial, but just out of principle and "that's cool" I'll make the update anyway :)

Courtesy of the OpenAI-empowered new Bing:
> String.AsSpan().IndexOf is faster than String.IndexOf in C# because it avoids memory allocation and garbage collection. Here is why:

> String.AsSpan() converts a string into a Span<char>[1](https://www.jitbit.com/alexblog/309-improving-c-performance-with-spant/)[2](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance)[3](https://www.codeproject.com/Articles/5269747/Using-Span-T-to-improve-performance-of-Csharp-code), which is a ref struct[4](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance)[2](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance) that can point to a region of memory without copying it[2](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance)[3](https://www.codeproject.com/Articles/5269747/Using-Span-T-to-improve-performance-of-Csharp-code)[5](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay)[6](https://learn.microsoft.com/en-us/dotnet/api/system.span-1?view=net-7.0). Span<char> can be allocated on the stack or inline in the containing type[2](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance)[5](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay)[6](https://learn.microsoft.com/en-us/dotnet/api/system.span-1?view=net-7.0), which means it does not require garbage collection[4](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance)[2](https://programmingfire.com/what-is-span-in-csharp-and-how-it-improves-the-performance).
Span<char>.IndexOf performs a non-copying slice operation on the span to find the index of a character or a substring[1](https://www.jitbit.com/alexblog/309-improving-c-performance-with-spant/)[7](https://www.dotnetperls.com/span)[5](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay). This means it does not create a new string object or allocate any memory on the heap[1](https://www.jitbit.com/alexblog/309-improving-c-performance-with-spant/)[7](https://www.dotnetperls.com/span)[5](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay).
String.IndexOf, on the other hand, performs a substring operation on the string to find the index of a character or a substring[8](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1846)[9](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1831)[10](https://stackoverflow.com/questions/51864673/c-sharp-readonlyspanchar-vs-substring-for-string-dissection). This means it creates a new string object and allocates memory on the heap[8](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1846)[9](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1831)[10](https://stackoverflow.com/questions/51864673/c-sharp-readonlyspanchar-vs-substring-for-string-dissection), which requires garbage collection later[8](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1846)[9](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1831)[10](https://stackoverflow.com/questions/51864673/c-sharp-readonlyspanchar-vs-substring-for-string-dissection).
Therefore, String.AsSpan().IndexOf is faster than String.IndexOf in C# because it avoids memory allocation and garbage collection by using Span<char>, which is a ref struct that can point to a region of memory without copying it.

Reply:
I think if you added StringComparison.Ordinal it would be similar but I do think span still has a tiny edge of that if I recall correctly.